### PR TITLE
#23023 Support Experiment into Url Map pages

### LINKED ISSUE #23023: Support Variant with URL Map
### Is your feature request related to a problem? Please describe.
When we try to render a URLMAP with a specific variant it is not working because the URLMapContent is render with the DEFAULT variant no matter if it had a version in the specific variant.

### Describe the solution you'd like
The contentlet should be render in the specific variant

### Additional context
here the Contentlet is returned with the DEFAULT variant version all the time

https://github.com/dotCMS/core/blob/3283e9732fe28b329cf3fd3633534a76a6f4806d/dotCMS/src/main/java/com/dotmarketing/cms/urlmap/URLMapAPIImpl.java#L63

### Acceptance criteria
- The contentlet should be render in the right version
- The redirect traffic must work with a URL Map
- Integration test

### PR DESCRIPTION
**Pull request by freddyucv**
We need to support Experiment using a detail-page, for this we need the follow:

- Take account the Url Map pattern to check when a redirect is needed.
- Render a detail-page inside a Variant it means using the right contentlet version.

### Proposed Changes
To Render:

* Take account the Variant version of the Contentlet when try to render a UrlMap Url

https://github.com/dotCMS/core/pull/24569/files#diff-f0cd9c5063627175a54b89cfb4d8c7705f3c9e385ddf8d831ad672ab334d0034R293

https://github.com/dotCMS/core/pull/24569/files#diff-3e34bcc38231d5f5c905331e68d93d63940b6f5e6276f67057d06ca76752651fR475

To Redirect:

* Use Regex to realize when a redirect is needed

https://github.com/dotCMS/core/pull/24569/files#diff-9417bef73d32b061e75306b404886f1b00b867e22696a5ba0b0717473190c922R21

* Calculate the Regex according to the Page's Uri

https://github.com/dotCMS/core/pull/24569/files#diff-df4c3ee11b473e0f4e91d5be7a6107bf6cdb4ab40bf0661a182dd298e00e5602R135

Also it fix this bug: https://github.com/dotCMS/core/issues/24511

* Add into the shouldBeInclude End point Response the redirect Regex

https://github.com/dotCMS/core/pull/24569/files#diff-df4c3ee11b473e0f4e91d5be7a6107bf6cdb4ab40bf0661a182dd298e00e5602R128

* Create a method to get the UrlMap Pattern from a page's id

API: https://github.com/dotCMS/core/pull/24569/files#diff-98406cfffe459aa06e1bbf615886de837936c80d7965159710391c9d97aa2742R169

Factory: https://github.com/dotCMS/core/pull/24569/files#diff-7acee0da7f6cdcef1bce71b548c9f8aa257c2308d58a639e2f146c8ff141ac60R92

* Additional change: When replace into the Experiment Js/Html code we was using replaceAll, now I am changing it to indexOf 

https://github.com/dotCMS/core/pull/24569/files#diff-df4c3ee11b473e0f4e91d5be7a6107bf6cdb4ab40bf0661a182dd298e00e5602R260

### CODE REVIEW

**Review by victoralfaro-dotcms — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeAPIImplTest.java [hunk 1]
```diff
@@ -2388,4 +2390,62 @@ public String toString() {
...
+		final List<String> urlMapped = APILocator.getContentTypeAPI(APILocator.systemUser())
+				.findUrlMappedPattern(htmlPageAsset.getIdentifier());
+
+		Assert.assertTrue(urlMapped.isEmpty());
+	}
+
+	/**
+	 * Method to test: {@link ContentTypeAPIImpl#findUrlMappedPattern(String)}
+	 * when: Called the method with Null
+	 * should: Throw a {@link NullPointerException}
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeAPIImplTest.java:2443]: NullPointer or IllegalArgument?

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeFactoryImplTest.java [hunk 2]
```diff
@@ -627,15 +635,66 @@ public void test_UpdateContentType_SetSortOrder() throws Exception {
...
+		final List<String> urlMapped = FactoryLocator.getContentTypeFactory()
+				.findUrlMappedPattern(htmlPageAsset.getIdentifier());
+
+		assertTrue(urlMapped.isEmpty());
+	}
+
+	/**
+	 * Method to test: {@link ContentTypeAPIImpl#findUrlMapped(String)}
+	 * when: Called the method with Null
+	 * should: Throw a {@link NullPointerException}
```
Comment (dsilvam) [inline dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeFactoryImplTest.java:693]: IllegalArgument

**Review by dsilvam — APPROVED** [review]

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeAPIImplTest.java (hunk 1 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeAPIImplTest.java:2443]: done https://github.com/dotCMS/core/pull/24569/commits/3000c2b26cc7b2e7f1752157f8a6d8642294b8bb#diff-4b70d46224e19c12838b1409554285a6b865eea342a8291c5735173a41d4acafR693

File: dotCMS/src/integration-test/java/com/dotcms/experiments/business/web/ExperimentWebAPIImplIT.java [hunk 3]
```diff
@@ -77,6 +85,253 @@ public void isUserIncluded() throws DotDataException, DotSecurityException {
...
+            ExperimentDataGen.end(experiment);
+        }
+    }
+
+    /**
+     * Method to test: {@link ExperimentWebAPIImpl#isUserIncluded(HttpServletRequest, HttpServletResponse, List)}
+     * When: You have one experiment running with 100% of traffic allocation and without targeting
+     * and using a page with uri equals to /index.
+     * Should: Return this experiment all the time and the redirect pattern should be equals to
+     * <code>^.*\/blog(\/index|\/)?(\?.*)?$</>
```
[the remark this answers is not in the document]
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/experiments/business/web/ExperimentWebAPIImplIT.java]: done https://github.com/dotCMS/core/pull/24569/commits/3000c2b26cc7b2e7f1752157f8a6d8642294b8bb#diff-7c1d4e7a1777c7c52ce86a8e605b66ccc8b79b6297d3c9e3f3d33059226ea3b3R170

File: dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeFactoryImplTest.java (hunk 2 above)
Comment (freddyucv) [reply dotCMS/src/integration-test/java/com/dotcms/contenttype/test/ContentTypeFactoryImplTest.java]: done https://github.com/dotCMS/core/pull/24569/commits/3000c2b26cc7b2e7f1752157f8a6d8642294b8bb#diff-4b70d46224e19c12838b1409554285a6b865eea342a8291c5735173a41d4acafR693

**Pull request merged**